models/issues/review.go

```diff
@@ -303,15 +309,38 @@ func CreateReview(ctx context.Context, opts CreateReviewOptions) (*Review, error
 		CommitID:     opts.CommitID,
 		Stale:        opts.Stale,
 	}
+
 	if opts.Reviewer != nil {
+		review.Type = opts.Type
 		review.ReviewerID = opts.Reviewer.ID
-	} else {
-		if review.Type != ReviewTypeRequest {
-			review.Type = ReviewTypeRequest
+
+		reviewCond := builder.Eq{"reviewer_id": opts.Reviewer.ID, "issue_id": opts.Issue.ID}
+		// make sure user review requests are cleared
+		if opts.Type != ReviewTypePending {
+			if _, err := sess.Where(reviewCond.And(builder.Eq{"type": ReviewTypeRequest})).Delete(new(Review)); err != nil {
+				return nil, err
+			}
 		}
+		// make sure if the created review gets dismissed no old review surface
+		if opts.Type == ReviewTypeApprove || opts.Type == ReviewTypeReject {
```

Comment:
What if "request review" again on a review which already has a "review request"?

Should the old review request be deleted or updated?

Reply:
well the upper level should prevent it ... but I can add that testcase and if it indeed is possible handle this case as workaround here ... the proper refactoring will be then done in #28544 ... :)

Reply:
@wxiaoguang added the testcase to the tests & extend the comments